File: Wire-iOS/Sources/UserInterface/InviteContacts/InviteContactsViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


@import WireExtensionComponents;

#import "InviteContactsViewController.h"
#import "InviteContactsViewController+Internal.h"
#import "ContactsViewController+Private.h"
#import "ContactsDataSource.h"
#import "ZClientViewController.h"
#import "WireSyncEngine+iOS.h"
#import "Wire-Swift.h"

@interface InviteContactsViewController () <ContactsViewControllerDelegate, ContactsViewControllerContentDelegate>
@end

@implementation InviteContactsViewController

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    
    if (self) {
        self.colorSchemeVariant = ColorSchemeVariantDark;
        self.delegate = self;
        self.contentDelegate = self;
        self.dataSource = [[ContactsDataSource alloc] init];
        self.dataSource.searchQuery = @"";
        
        self.title = [NSLocalizedString(@"contacts_ui.title", @"") uppercaseString];
    }
    
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setupStyle];
}

- (BOOL)sharingContactsRequired
{
    return YES;
}

- (void)inviteUserOrOpenConversation:(ZMSearchUser *)user fromView:(UIView *)view
{
    if (user.isConnected) {
        [[ZClientViewController sharedZClientViewController] selectConversation:user.user.oneToOneConversation
                                                                    focusOnView:YES
                                                                       animated:YES];
    } else if (user.user.isPendingApprovalBySelfUser && ! user.user.isIgnored) {
        [[ZClientViewController sharedZClientViewController] selectIncomingContactRequestsAndFocusOnView:YES];
    } else if (user.user.isPendingApprovalByOtherUser && ! user.user.isIgnored) {
        [[ZClientViewController sharedZClientViewController] selectConversation:user.user.oneToOneConversation
                                                                    focusOnView:YES
                                                                       animated:YES];
        
    } else if (user.user != nil && ! user.user.isIgnored && ! user.user.isPendingApprovalByOtherUser) {
        NSString *messageText = [NSString stringWithFormat:NSLocalizedString(@"missive.connection_request.default_message",@"Default connect message to be shown"), user.user.displayName, [ZMUser selfUser].name];
        
        [[ZMUserSession sharedSession] enqueueChanges:^{
            [user connectWithMessage:messageText];
        } completionHandler:^{
            [self.tableView reloadData];
        }];
    } else {
        UIAlertController * alertController = [self inviteContact:user.contact fromView:view];

        [alertController presentInNotificationsWindow];
    }
}

#pragma mark - ContactsViewControllerDelegate

- (void)contactsViewControllerDidCancel:(ContactsViewController *)controller
{
    [controller dismissViewControllerAnimated:YES completion:nil];
}

- (void)contactsViewControllerDidNotShareContacts:(ContactsViewController *)controller
{
    [controller dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - ContactsViewControllerContentDelegate

- (BOOL)contactsViewController:(ContactsViewController *)controller shouldDisplayActionButtonForUser:(ZMSearchUser *)user
{
    return YES;
}

- (NSArray *)actionButtonTitlesForContactsViewController:(ContactsViewController *)controller
{
    return @[
             NSLocalizedString(@"contacts_ui.action_button.open", @""),
             NSLocalizedString(@"contacts_ui.action_button.invite", @""),
             NSLocalizedString(@"connection_request.send_button_title", @""), // TODO: add separate string contacts_ui.action_button.connect
             ];
}

- (NSUInteger)contactsViewController:(ContactsViewController *)controller actionButtonTitleIndexForUser:(ZMSearchUser *)user
{
    if (user.isConnected || ((user.user.isPendingApprovalByOtherUser || user.user.isPendingApprovalBySelfUser) && ! user.user.isIgnored)) {
        return 0;
    } else if (user.user != nil && ! user.user.isIgnored && ! user.user.isPendingApprovalByOtherUser) {
        return 2;
    }
    else {
        return 1;
    }
}

- (void)contactsViewController:(ContactsViewController *)controller actionButton:(UIButton *)actionButton pressedForUser:(ZMSearchUser *)user
{
    [self inviteUser:user fromView:actionButton];
}

- (void)contactsViewController:(ContactsViewController *)controller didSelectCell:(ContactsCell *)cell forUser:(ZMSearchUser *)user
{
    [self inviteUser:user fromView:cell];
}

@end

File: Wire-iOS/Sources/UserInterface/InviteContacts/InviteContactsViewController.swift
//
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation

final class InviteContactsViewController: ContactsViewController {
    init() {
        super.init(nibName: nil, bundle: nil)
        
        delegate = self
        contentDelegate = self
        dataSource = ContactsDataSource()
        dataSource?.searchQuery = ""
        
        title = "contacts_ui.title".localized.uppercased()
        
        setupStyle()
    }
    
    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    override func sharingContactsRequired() -> Bool {
        return true
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        ///hide titleLabel and cancel cross button, which is duplicated in the navi bar
        
        let subViewConstraints = [titleLabelHeightConstraint, titleLabelTopConstraint, titleLabelBottomConstraint, closeButtonTopConstraint, closeButtonBottomConstraint, searchHeaderTopConstraint]
        
        if navigationController != nil {
            titleLabel.isHidden = true
            
            cancelButton.isHidden = true
            closeButtonHeightConstraint.constant = 0
            subViewConstraints.forEach(){ $0.isActive = false }
            
            topContainerHeightConstraint.isActive = true
            searchHeaderWithNavigatorBarTopConstraint.isActive = true
        } else {
            titleLabel.isHidden = false
            
            cancelButton.isHidden = false
            
            closeButtonHeightConstraint.constant = 16
            topContainerHeightConstraint.isActive = false
            searchHeaderWithNavigatorBarTopConstraint.isActive = false
            
            subViewConstraints.forEach(){ $0.isActive = true }
        }
        
        view.layoutIfNeeded()
    }
    
    override func setupStyle() {
        super.setupStyle()
        
        view.backgroundColor = .clear
        
        tableView.backgroundColor = .clear
        tableView.separatorStyle = .none
        tableView.sectionIndexBackgroundColor = .clear
        tableView.sectionIndexColor = .accent()
        
        bottomContainerSeparatorView.backgroundColor = UIColor.from(scheme: .separator, variant: .dark)
        bottomContainerView.backgroundColor = UIColor.from(scheme: .searchBarBackground, variant: .dark)
        
        titleLabel.textColor = UIColor.from(scheme: .textForeground, variant: .dark)
    }
    
    private func invite(user: ZMSearchUser, from view: UIView) {
        
        // Prevent the overlapped visual artifact when opening a conversation
        if let navigationController = self.navigationController, self == navigationController.topViewController && navigationController.viewControllers.count >= 2 {
            navigationController.popToRootViewController(animated: false) {
                self.inviteUserOrOpenConversation(user, from:view)
            }
        } else {
            inviteUserOrOpenConversation(user, from:view)
        }
    }
    
    private func inviteUserOrOpenConversation(_ user: ZMSearchUser, from view: UIView) {
        let searchUser: ZMUser? = user.user
        let isIgnored: Bool? = searchUser?.isIgnored
        
        let selectOneToOneConversation: Completion = {
            if let oneToOneConversation = searchUser?.oneToOneConversation {
                ZClientViewController.shared?.select(conversation: oneToOneConversation, focusOnView: true, animated: true)
            }
        }
        
        if user.isConnected {
            selectOneToOneConversation()
        } else if searchUser?.isPendingApprovalBySelfUser == true &&
            isIgnored == false {
            ZClientViewController.shared?.selectIncomingContactRequestsAndFocus(onView: true)
        } else if searchUser?.isPendingApprovalByOtherUser == true &&
            isIgnored == false {
            selectOneToOneConversation()
        } else if let unwrappedSearchUser = searchUser,
                !unwrappedSearchUser.isIgnored &&
                !unwrappedSearchUser.isPendingApprovalByOtherUser {
            let displayName = unwrappedSearchUser.displayName
            let messageText = String(format: "missive.connection_request.default_message".localized, displayName, ZMUser.selfUser().name ?? "")
            
            ZMUserSession.shared()?.enqueueChanges({
                user.connect(message: messageText)
            }, completionHandler: {
                self.tableView.reloadData()
            })
        } else if let contact = user.contact,
                  let alertController = invite(contact, from: view) {
            
            
            AppDelegate.shared.window?.rootViewController?.present(alertController, animated: true)
        }
    }
}

extension InviteContactsViewController: ContactsViewControllerContentDelegate {
    
    func contactsViewController(_ controller: ContactsViewController, shouldSelect user: ZMSearchUser) -> Bool {
        return true
    }

    var shouldDisplayActionButton: Bool {
        return true
    }
    
    func actionButtonTitles(for controller: ContactsViewController) -> [String] {
        return ["contacts_ui.action_button.open".localized,
                "contacts_ui.action_button.invite".localized,
                "connection_request.send_button_title".localized]
    }
    
    func contactsViewController(_ controller: ContactsViewController,
                                actionButtonTitleIndexFor user: UserType?,
                                isIgnored: Bool) -> Int {
        guard let user = user else { return 1 }
        
        if user.isConnected ||
           ((user.isPendingApprovalByOtherUser ||
             user.isPendingApprovalBySelfUser) && isIgnored) {
            return 0
        } else if !isIgnored,
            !user.isPendingApprovalByOtherUser {
            return 2
        }
        
        return 1
    }
    
    func contactsViewController(_ controller: ContactsViewController, actionButton: UIButton, pressedFor user: ZMSearchUser) {
        invite(user: user, from: actionButton)
    }
    
    func contactsViewController(_ controller: ContactsViewController, didSelect cell: ContactsCell, for user: ZMSearchUser) {
        invite(user: user, from: cell)
    }
}

extension InviteContactsViewController: ContactsViewControllerDelegate {
    func contactsViewControllerDidCancel(_ controller: ContactsViewController) {
        controller.dismiss(animated: true)
    }
    
    func contactsViewControllerDidNotShareContacts(_ controller: ContactsViewController) {
        controller.dismiss(animated: true)
    }

    func contactsViewControllerDidConfirmSelection(_ controller: ContactsViewController) {
        //no-op
    }
}
